### PR TITLE
Rework API

### LINKED ISSUE #4321: API create mailbox fails for subdomains
- [ X] I understand that not following or deleting the below instructions will result in immediate closure and/or deletion of my issue.
- [ X] I have understood that this bug report is dedicated for bugs, and not for support-related inquiries.
- [ X] I have understood that answers are voluntary and community-driven, and not commercial support.
- [ X] I have verified that my issue has not been already answered in the past. I also checked previous [issues](https://github.com/mailcow/mailcow-dockerized/issues).

## Summary
When calling the API to add mailboxes, it quietly refuses to create emails linked to a subdomain.
e.g.
john@domain.com --> works
john@demo.domain.com --> does NOT work

I can manually create john@demo.domain.com in the UI but not via API

## Logs
No errors in logs

## Reproduction
Add a domain and its subdomain, attempt to create emails programmatically

## System information

| Question | Answer |
| --- | --- |
| My operating system |Ubuntu 20.04.3 LTS |
| Is Apparmor, SELinux or similar active? | No |
| Virtualization technlogy | No |
| Server/VM specifications (Memory, CPU Cores) | 8MB 4Cores |
| Docker Version (`docker version`) | 20.10.10 |
| Docker-Compose Version (`docker-compose version`) | 1.29.2 |
| Reverse proxy (custom solution) | None |

- Output of `git diff origin/master`, any other changes to the code? No changes to master
- All third-party firewalls and custom iptables rules are unsupported. No custom rules
- DNS problems? No DNS issues -- 151.101.65.69 151.101.1.69 151.101.129.69 151.101.193.69

### LINKED ISSUE #4349: API returns wrong status code
## Summary

The Mailcow API should throw 404 if something is not found.
Instead it gives a HTTP 200 OK status code.

## Reproduction

Try to get a non existing domain via the API.
image

(Cleaned up the issue, to make it easier readable). 

That's not really a bug, more a missing feature.

### PR DESCRIPTION
**Pull request by Lulalaby**
## Description
This PR aims to rework the API to give better and more consistence responses.
This is WIP.
Creating this PR already for transparency and to give maintainers/contributors the possibility to contribute.

## This PR closes the following Issues
Closes #4349 
Fixes #4321

### CODE REVIEW

**Comment by Lulalaby** [comment]
Furthermore we check on #4321 and try to fix it.

**Comment by Lulalaby** [comment]
As note: At the actual state libs/clients would've to do this workaround:
https://github.com/Aiko-IT-Systems/DisCatSharp.Mailcow/blob/bd96174c07a42faeca9791bee14458df26f0ef73/DisCatSharp.Mailcow/Rest/MailcowApiClient.cs#L87-L110

File: data/web/json_api.php [hunk 1]
```diff
@@ -1,1772 +1,1831 @@
...
+      break;
+    case 'error':
+      $lvlint = 4;
+      break;
+    default:
+      $lvlint = 0;
+      break;
+  }
+  return $lvlint;
+}
```
Comment (Lulalaby) [inline data/web/json_api.php:1831]: What are the thoughts about returning an int for the type on messages?
That would make it easier to work with enums in other clients / languages.

**Comment by Lulalaby** [comment]
Added data/web/json_api_v2.php and created a route to /api/v2 to avoid breaking the api v1

**Comment by Lulalaby** [comment]
I'm a bit ill atm and still not done. Sorry for that

**Comment by feldsam** [comment]
Hello, would be fine to do more work on refactoring. There are lot of bad patterns in original api and you just copy it to new file. So many switch/case, repeating and not clear code and so on.

**Comment by Lulalaby** [comment]
Yeah I would love to elaborate on this.
I've tried to fix it up but got busy with private stuff, so haven't had time yet. I think in generally the whole else if switch statements are kinda.. bad practice here.

**Comment by Lulalaby** [comment]
Optimal it would be if we can implement a more advanced api key system to allow btw read access per route and so on.
But that's nothing I can do alone.
With the past changes I tried to get familiar how mailcow and its api works, but to do a full rewrite I need time and help.

**Comment by feldsam** [comment]
> But that's nothing I can do alone.
> With the past changes I tried to get familiar how mailcow and its api works, but to do a full rewrite I need time and help.

Yes, I understand, I am in the similar situation.

**Comment by Lulalaby** [comment]
I'll close this for now, and make a new one or reopen this when I got time again for this.
I still want to work on that! Just that you all now.

**Pull request closed without merging**